### ISSUE TITLE
Handling of missing values in categorical encoders

### ISSUE THREAD

**Issue by david-cortes** [feature-engine/feature_engine#481]
Some transformers such as the `CountFrequency` categorical encoder have options for handling new unseen categories during calls to `transform`. In some cases, unseen categories and explicit missing values might be getting mixed together however.

It would be good to have some distinction between the two and potentially different options for courses of action that would be consistent across transformers. For example, if `CountFrequency` encoder receives an explicit missing value during transform, the user might want to do one of the following:
* Propagate those missing values to the output of `transform`.
* Treat them the same as unseen categories (encoding them with zeros).
* Encode them with the frequency of missing values that were passed to the training data (treating missing values as a separate category).
* Raise an error.

In most of the cases, only the last alternative (raising an error) is supported, but would be ideal to have options and to make explicit in the documentation what happens with missing values.

**Comment by david-cortes** [comment]
Many transformers actually fail to fit when the data has missing values, even though their presence shouldn't affect the calculations that the transformer does. One might want to preserve missing values as missing in the output before passing them to some further transformer for example (such as a multivariate imputer).

**Comment by solegalli** [comment]
All encoders will fail if the input has missing data. They will fail during fit and they will also fail during transform. This is intentional.

The user can impute missing data beforehand, for example with the CategoricalImputer() and replace nan by a string, for example "Missing", if they want to preserve that as a separate category.

Unseen categories are a different story. They will only be taken care of during transform in 1 of 2 ways for most transformers: either fail or return nan in place of the unseen category. The countfrequency now has the option to return 0.

I am not sure I understand what the issue is.

Maybe you could give a code example?

**Comment by david-cortes** [comment]
Example use cases for passing explicit missing values:
* I have a data frame with mixed columns and missing values in many of them. I want to fit a regression model using a software that handles missing values without imputation, such as lightgbm. Some categorical columns have high cardinality and lightgbm doesn't handle them well, so I want to encode them with some transformers.
* I have a data frame with mixed columns and missing values in some columns. I want to fit a linear model, for which I want to impute the missing values. I again have high-cardinality categorical features and I know that I can get better results by using a multivariate imputer with the transformed columns when I impute numerical columns for example.
* I might want to combine the results of the transformations across different columns and fill missing values using those results. If they get imputed beforehand, the end result will be different.

**Comment by david-cortes** [comment]
In the case of lightgbm, when it encounters NaNs in a categorical variable during fitting, at each node that makes a split on that variable, it decides to which branch to assign NaNs by checking on which branch they would decrease the loss function the most, which is something that could not be achieved by using the same imputed value at every tree and node - hence why one would want to pass NaNs as-is instead of imputing them.
